java/google/registry/rde/XjcToDomainResourceConverter.java

```diff
@@ -0,0 +1,329 @@
+        .build();
+    ofy().transact(new VoidWork() {
+
+      @Override
+      public void vrun() {
+        ofy().save().entity(historyEntry);
+      }
+    });
+    return historyEntry;
+  }
+
+  /** Returns {@link Key} for registrant from foreign key */
+  private static Key<ContactResource> convertRegistrant(String contactId) {
+    Key<ContactResource> key = ForeignKeyIndex.loadAndGetKey(ContactResource.class, contactId, DateTime.now());
+    checkState(key != null, "Registrant not found: '%s'", contactId);
+    return key;
+  }
+
+  /** Converts {@link XjcDomainNsType} to {@link ImmutableSet<Key<HostResource>>}. */
+  private static ImmutableSet<Key<HostResource>> convertNameservers(XjcDomainNsType ns) {
+    // If no hosts are specified, return an empty set
+    if (ns == null || (ns.getHostAttrs() == null && ns.getHostObjs() == null)) {
+      return ImmutableSet.of();
+    }
+    // Domain linked hosts may be specified by host object or attribute, but not both.
+    if (ns.getHostAttrs() == null || ns.getHostAttrs().isEmpty()) {
+      // Specified as object refs
+      return ImmutableSet
+          .copyOf(Iterables.transform(ns.getHostObjs(), HOST_OBJ_CONVERTER));
+    } else {
+      // Specified as attributes
+      throw new IllegalArgumentException("Host attributes are not yet supported");
+    }
+  }
+
+  /** Converts {@link XjcRdeDomainTransferDataType} to {@link TransferData}. */
+  private static TransferData convertDomainTransferData(XjcRdeDomainTransferDataType data) {
+    if (data == null) {
+      return TransferData.EMPTY;
+    }
+    return new TransferData.Builder()
+        .setTransferStatus(TRANSFER_STATUS_MAPPER.xmlToEnum(data.getTrStatus().value()))
+        .setGainingClientId(data.getReRr().getValue())
+        .setLosingClientId(data.getAcRr().getValue())
+        .setTransferRequestTime(data.getReDate())
+        .setPendingTransferExpirationTime(data.getAcDate())
+        .build();
+  }
+
+  /** Converts {@link XjcDomainStatusType} to {@link StatusValue}. */
+  private static StatusValue convertStatusType(XjcDomainStatusType type) {
+    return StatusValue.fromXmlName(type.getS().value());
+  }
+
+  /** Converts {@link XjcSecdnsDsDataType} to {@link DelegationSignerData}. */
+  private static DelegationSignerData convertSecdnsDsDataType(XjcSecdnsDsDataType secdns) {
+    return DelegationSignerData.create(
+        secdns.getKeyTag(),
+        (int) secdns.getAlg(),
+        (int) secdns.getDigestType(),
+        secdns.getDigest()
+    );
+  }
+
+  /** Converts {@link XjcDomainContactType} to {@link DesignatedContact}. */
+  private static DesignatedContact convertContactType(XjcDomainContactType contact) {
+    String contactId = contact.getValue();
+    Key<ContactResource> key = ForeignKeyIndex.loadAndGetKey(ContactResource.class, contactId, DateTime.now());
+    checkState(key != null, String.format("Contact not found: '%s'", contactId));
```

Comment:
checkState has a varargs version that takes the conditional, a format string, and then varargs for the format objects.  No need to call String.format() separately.